File: vertx-elasticsearch/src/main/java/fr/myprysm/vertx/elasticsearch/action/support/ShardInfoFailure.java
package fr.myprysm.vertx.elasticsearch.action.support;

import io.vertx.codegen.annotations.DataObject;
import io.vertx.core.json.JsonObject;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;
import org.elasticsearch.rest.RestStatus;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Accessors(chain = true)
@EqualsAndHashCode(callSuper = true)
@DataObject(generateConverter = true)
public class ShardInfoFailure extends Failure {

    private String shardId;
    private String nodeId;
    private RestStatus status;
    private boolean primary;

    public ShardInfoFailure(String shardId, String nodeId, RestStatus status, boolean primary, String cause) {
        super(cause);
        this.shardId = shardId;
        this.nodeId = nodeId;
        this.status = status;
        this.primary = primary;
    }

    public ShardInfoFailure(JsonObject json) {
        super(json);
        ShardInfoFailureConverter.fromJson(json, this);
    }

    public JsonObject toJson() {
        JsonObject json = super.toJson();
        ShardInfoFailureConverter.toJson(this, json);
        return json;
    }
}

File: vertx-elasticsearch/src/main/kotlin/fr/myprysm/vertx/elasticsearch/kotlin/action/support/ShardInfoFailure.kt
package fr.myprysm.vertx.elasticsearch.kotlin.action.support

import fr.myprysm.vertx.elasticsearch.action.support.ShardInfoFailure
import org.elasticsearch.rest.RestStatus

fun ShardInfoFailure(
        cause: String? = null,
        nodeId: String? = null,
        primary: Boolean? = null,
        shardId: String? = null,
        status: RestStatus? = null): ShardInfoFailure = fr.myprysm.vertx.elasticsearch.action.support.ShardInfoFailure().apply {

    if (cause != null) {
        this.setCause(cause)
    }
    if (nodeId != null) {
        this.setNodeId(nodeId)
    }
    if (primary != null) {
        this.setPrimary(primary)
    }
    if (shardId != null) {
        this.setShardId(shardId)
    }
    if (status != null) {
        this.setStatus(status)
    }
}

